src/java/org/apache/cassandra/index/sai/plan/Plan.java

```diff
@@ -1459,8 +1622,24 @@ public TableMetrics(long rows, double avgCellsPerRow, double avgBytesPerRow, int
     public interface Executor
     {
         Iterator<? extends PrimaryKey> getKeysFromIndex(Expression predicate);
-        Iterator<? extends PrimaryKey> getTopKRows(RowFilter.Expression ordering);
-        Iterator<? extends PrimaryKey> getTopKRows(RangeIterator keys, RowFilter.Expression ordering);
+        Iterator<? extends PrimaryKey> getTopKRows(RowFilter.Expression ordering, int softLimit);
+        Iterator<? extends PrimaryKey> getTopKRows(RangeIterator keys, RowFilter.Expression ordering, int softLimit);
+    }
+
+    /**
+     * Outsources more complex cost estimates to external components.
+     * Some components may collect stats on previous data execution and deliver more accurate estimates based
+     * on that state.
+     */
+    public interface CostEstimator
```

Comment:
@michaeljmarshall flagging this as another piece of currently-ann-only code that probably needs to be generalized for full `order by`